sa/sa.go

```diff
@@ -829,39 +878,77 @@ func (ssa *SQLStorageAuthority) RevokeCertificate(ctx context.Context, req *sapb
+	_, overallError := db.WithTransaction(ctx, ssa.dbMap, func(tx db.Executor) (interface{}, error) {
+		thisUpdate := time.Unix(0, req.DateNS)
+		revokedDate := time.Unix(0, req.BackdateNS)
 
-	res, err := ssa.dbMap.ExecContext(ctx,
-		`UPDATE certificateStatus SET
-				revokedReason = ?,
-				ocspLastUpdated = ?
-			WHERE serial = ? AND status = ? AND revokedReason != ? AND revokedDate = ?`,
-		revocation.Reason(ocsp.KeyCompromise),
-		thisUpdate,
-		req.Serial,
-		string(core.OCSPStatusRevoked),
-		revocation.Reason(ocsp.KeyCompromise),
-		revokedDate,
-	)
-	if err != nil {
-		return nil, err
-	}
-	rows, err := res.RowsAffected()
-	if err != nil {
-		return nil, err
-	}
-	if rows == 0 {
-		// InternalServerError because we expected this certificate status to exist,
-		// to already be revoked for a different reason, and to have a matching date.
-		return nil, berrors.InternalServerError("no certificate with serial %s and revoked reason other than keyCompromise", req.Serial)
+		res, err := tx.ExecContext(ctx,
+			`UPDATE certificateStatus SET
+					revokedReason = ?,
+					ocspLastUpdated = ?
+				WHERE serial = ? AND status = ? AND revokedReason != ? AND revokedDate = ?`,
+			revocation.Reason(ocsp.KeyCompromise),
+			thisUpdate,
+			req.Serial,
+			string(core.OCSPStatusRevoked),
+			revocation.Reason(ocsp.KeyCompromise),
+			revokedDate,
+		)
+		if err != nil {
+			return nil, err
+		}
+		rows, err := res.RowsAffected()
+		if err != nil {
+			return nil, err
+		}
+		if rows == 0 {
+			// InternalServerError because we expected this certificate status to exist,
+			// to already be revoked for a different reason, and to have a matching date.
+			return nil, berrors.InternalServerError("no certificate with serial %s and revoked reason other than keyCompromise", req.Serial)
+		}
+
+		if req.ShardIdx != 0 {
+			var rcm revokedCertModel
+			// Note: this query MUST be updated to enforce the same preconditions as
+			// the "UPDATE certificateStatus SET revokedReason..." above if this
+			// query ever becomes the first or only query in this transaction. We are
+			// currently relying on the query above to exit early if the certificate
+			// does not have an appropriate status.
+			err = tx.SelectOne(
+				ctx, &rcm, `SELECT * FROM revokedCertificates WHERE serial = ?`, req.Serial)
+			if db.IsNoRows(err) {
+				// TODO: Remove this fallback codepath once we know that all unexpired
+				// certs marked as revoked in the certificateStatus table have
+				// corresponding rows in the revokedCertificates table. That should be
+				// 90+ days after the RA starts sending ShardIdx in its
+				// RevokeCertificateRequest messages.
+				err = addRevokedCertificate(ctx, tx, req, revokedDate)
+				if err != nil {
+					return nil, err
+				}
+				return nil, nil
+			} else if err != nil {
+				return nil, fmt.Errorf("retrieving revoked certificate row: %w", err)
+			}
+
+			rcm.RevokedReason = revocation.Reason(ocsp.KeyCompromise)
```

Comment:
Don't we also need to update the revokedDate in this case?

Reply:
The existing code doesn't update the revockedDate in the certificateStatus table. The reason is that per the BRs we're only allowed to move the revocation timestamp backwards (in case someone learns that the key was actually compromised earlier than the cert was originally revoked) but ACME revocation requests don't come with a date, they just use the current date. So we don't allow updating revocation to change the revokedDate.